horovod/tensorflow/mpi_ops.cc

```diff
@@ -163,6 +163,9 @@ class TFOpContext : public common::OpContext {
   AllocateOutput(common::TensorShape shape,
                  std::shared_ptr<common::Tensor>* tensor) override;
   virtual common::Status
+  AllocateOutput(int output_index, common::TensorShape shape,
+                 std::shared_ptr<common::Tensor>* tensor) override;
+  virtual common::Status
```

Comment:
Instead of defining an overloaded `AllocateOutput` with the additional `output_index` argument, it might be cleaner to just add an `output_index` to the existing method with a default value of 0.

Reply:
That was my first idea. But default arguments on virtual and override functions seem to be frowned upon in C++, which I learnt from the Clang warnings I was prompted with (see e.g. https://google.github.io/styleguide/cppguide.html#Default_Arguments). Realistically, there probably is only the potential of issues if any derived class were to specify a different default value from the base class, which is quite unlikely in this case. So if you prefer, I could combine these to a single definition with a default argument (and ignore warnings).

Reply:
Ok, sounds like a good enough reason to me to do things the way you did. Thanks for being thorough!